src/mlpack/methods/ann/layer/lookup_impl.hpp

```diff
@@ -54,17 +63,34 @@ void Lookup<InputDataType, OutputDataType>::Gradient(
     const arma::Mat<eT>& error,
     arma::Mat<eT>& gradient)
 {
-  gradient = arma::zeros<arma::Mat<eT> >(weights.n_rows, weights.n_cols);
-  gradient.cols(arma::conv_to<arma::uvec>::from(input) - 1) = error;
+  const size_t seqLength = input.n_rows;
+  const size_t batchSize = input.n_cols;
+
+  arma::Cube<eT> errorTemp(const_cast<arma::Mat<eT>&>(error).memptr(),
+      embeddingSize, seqLength, batchSize, false, false);
+
+  gradient.set_size(arma::size(weights));
+  gradient.zeros();
+
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    gradient.cols(arma::conv_to<arma::uvec>::from(input.col(i)) - 1)
+        += errorTemp.slice(i);
+  }
```

Comment:
@lozhnikov: I think we will need to make some changes here. As you suggested the shape of input to be `(tgtSeqLen, embedDim, batchSize)` and not `(embedDim, tgtSeqLen, batchSize)` in #2375, here also I think it should be `(seqLength, embeddingSize, batchSize)`. Otherwise, if Lookup is used before multihead attention, the dimensions will not be compatible.